### ISSUE TITLE
GKC to Current Attachments may process a piece more than once

### ISSUE THREAD

**Issue by riverwanderer** [vassalengine/vassal#12433]
Should probably not be the case ?

Testing in Vassal 3.7beta4.

**Comment by BrentEaston** [comment]
I guess the question is, how do we reproduce the double attachment? Do you have a module and save game in a state where the next GKC will have this problem?

**Comment by BrentEaston** [comment]
I'm so glad someone is testing this stuff! Your help is invaluable!!!!

**Comment by riverwanderer** [comment]
> I'm so glad someone is testing this stuff! Your help is invaluable!!!!

Very happy to contribute. In this case, Attachments have turbo-charged some re-development I had just started. I am using them extensively in new and re-worked features of the C&C Nap module. Even so, I am scratching the surface. It will be good to see other module developers using the feature. Once its bedded in, I hope to write an article on the forum to highlight how I've used it.

**Comment by riverwanderer** [comment]
> I guess the question is, how do we reproduce the double attachment? Do you have a module and save game in a state where the next GKC will have this problem?

This demo module will walk through the issue. Open "New Test" from the file menu and follow the instructions.

[Demo GKC Current Attachments.zip](https://github.com/vassalengine/vassal/files/11956239/Demo.GKC.Current.Attachments.zip)

**Comment by BrentEaston** [comment]
Not sure what I'm looking at. I can't see anything called 'New Test' on the file menu and when I look at that module, I don't see any pieces with Attachment traits. Are you sure this is the right one?

**Comment by riverwanderer** [comment]
> Not sure what I'm looking at. I can't see anything called 'New Test' on the file menu and when I look at that module, I don't see any pieces with Attachment traits. Are you sure this is the right one?

@BrentEaston  Puzzled. I just downloaded that zip file and I see what I expect (includng 'New Test')

<img width="414" alt="image" src="https://github.com/vassalengine/vassal/assets/58135975/ed261d84-e7a4-43c4-b3b7-d135fd580e99">

**Comment by BrentEaston** [comment]
I see 'New Game' in that spot on the menu. When I select it, I get a new game thus:

![image](https://github.com/vassalengine/vassal/assets/1622667/95681147-cf4c-4b20-9724-29d978cc5226)

**Comment by riverwanderer** [comment]
oh, poop.  Weird I never saw that. Maybe it's the wrong version. Looking now...

**Comment by riverwanderer** [comment]
Hang on, that is the wrong module title.... are you sure you are running the right thing? What's in your Vassal library?

This is what you should see and what I see when I do New Test from the module I downloaded here:
<img width="635" alt="image" src="https://github.com/vassalengine/vassal/assets/58135975/5764c97a-d8fe-4c7a-bd02-e5ba51937d35">

**Comment by BrentEaston** [comment]
I know what has happened.

Inside the zip file you posted is another vmod file as well as buildfile etc. I'm running that vmod that is inside your module. I didn't realise the zip is the module, OI though you has just accidentally included some junk.

**Comment by BrentEaston** [comment]
Ok, got something happening now. Thanks.

**Comment by BrentEaston** [comment]
Ok, The issue is that the pieces are in fact attached more than once, on two different attachments. I think it is reasonable to expect that in this case, the Current attachments fastmatch will only match each piece tested a maximum of once, even if it satisifies the test multiple times. I can fix this fairly easily,

Now, I remember you discussing about Fast Matching on a specifically named attachment only. I can see the need for this now. Sending a GKC to all attachments is too scatter-gun.

There's a couple of ways to handle the Fastmatch Ui.

1. Leave the Current Attachments as is and add a `Named Attachment` which will open up a box and let you enter an Attachment Name (or expression) and only that attachment will be matched.

2. Have the Current Attachments option open up an 'Attachment Name' box, with a hint text of 'Leave blank for all'.  Expressions allowed.

**Comment by BrentEaston** [comment]
I'm working on option 2

**Comment by riverwanderer** [comment]
> There's a couple of ways to handle the Fastmatch Ui.
> 
> 1. Leave the Current Attachments as is and add a `Named Attachment` which will open up a box and let you enter an Attachment Name (or expression) and only that attachment will be matched.
> 2. Have the Current Attachments option open up an 'Attachment Name' box, with a hint text of 'Leave blank for all'.  Expressions allowed.

I might lean towards (1) for consistency with the Current/Specific of the other options, but that's a fussy point compared to the overall solution.

I agree on capping GKC hits at one per piece. That is consistent, iirc, with @Cattlesquat's fix that he did for issue #12116.

**Comment by BrentEaston** [comment]
Try the build 12433-AttachmentFMIssues, it contains the combined fixed for 12433 & 12444.

**Comment by riverwanderer** [comment]
@BrentEaston - I tested the named Attachment feature of #12448 and it works. Much more straightforward / compact to use than having to filter using the other fields. I'll post a short thread on discord showing how I'm using this, just for info.

Though the new Current Attachments name field means I can specify just one Attachment, I also tested with a blank field. That shows mean that your change to ensure 1 x GKC hit per piece has also worked.

Lastly, I adapted Attachment & GKC configuration to demonstrate that "Within a Deck, apply to"... setting is respected and it passed that test.

Re #12444 - I am not quite sure how PR #12448 directly addresses that point, though I can see that it makes more workarounds available/easier. Auto attachment will, I think, still not be as useful as it could be, as discussed in #12444. Maybe I've missed something ?
